core/src/main/scala/org/http4s/Message.scala

```diff
@@ -294,13 +294,19 @@ sealed abstract case class Request[F[_]](
       attributes = attributes
     )
 
-  lazy val (scriptName, pathInfo) = {
-    val caret = attributes.lookup(Request.Keys.PathInfoCaret).getOrElse(0)
+  lazy val (scriptName, pathInfo) =
     uri.path.splitAt(caret)
-  }
 
-  def withPathInfo(pi: String): Self =
-    withUri(uri.withPath(scriptName + pi))
+  private def caret =
+    attributes.lookup(Request.Keys.PathInfoCaret).getOrElse(0)
+
+  private def withCaret(caret: Int) =
+    withAttribute(Request.Keys.PathInfoCaret, caret)
+
+  def withPathInfo(pi: String): Self = {
+    val oldCaret = caret
+    withUri(uri.withPath(scriptName + pi)).withCaret(oldCaret)
```

Comment:
Instead of having to lookup the caret, couldn't you just do:
```
copy(uri = uri.withPath(scriptName + pi))
```

Reply:
:+1: That's a good point.  `withUri` clears the caret, but this can use `copy` directly.  I changed it to this.